ultravox/data/datasets.py

```diff
@@ -1167,3 +1176,8 @@ def __iter__(self):
             if self._num_samples is not None and i >= self._num_samples:
                 break
             yield sample
+
+    def __len__(self):
+        return (
+            self._num_samples if self._num_samples is not None else len(self._dataset)
```

Comment:
We may need to add the following in `__init__`, otherwise, `__len__` is not correct if `num_samples` is larger than the size of the dataset. 

```
self._num_samples = min(num_samples, len(self._dataset))

Reply:
I think we probably need to include the actual data length somewhere in the config, or else we have to iterate through the datset to get the length.

Reply:
This is the fundamental problem with using epochs and streaming, AFAICT HF doesn't know the dataset length up front (even if you look at the info in the dataset card, it might be wrong). For datasets that we rely on I suppose we could add a hardcoded length in our wrapper, but this will naturally be somewhat fragile.

Reply:
I did some searching on this topic, and it looks like there’s no way to get around this with streaming mode. Right now, I rely on the info in the dataset card/readme file to find the number of rows. Maybe we just have to live with it.